File: src/main/java/ru/citeck/ecos/actions/codeinsight/GenerateEcosFormComponent.java
package ru.citeck.ecos.actions.codeinsight;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.intellij.codeInsight.actions.SimpleCodeInsightAction;
import com.intellij.ide.actions.QualifiedNameProviderUtil;
import com.intellij.json.psi.JsonArray;
import com.intellij.json.psi.JsonFile;
import com.intellij.json.psi.JsonObject;
import com.intellij.json.psi.JsonProperty;
import com.intellij.openapi.actionSystem.Presentation;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.command.CommandProcessor;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.project.DumbService;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.ui.popup.JBPopupFactory;
import com.intellij.openapi.wm.WindowManager;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import com.intellij.psi.util.PsiTreeUtil;
import org.jetbrains.annotations.NotNull;
import ru.citeck.ecos.ServiceRegistry;
import ru.citeck.ecos.codeinsight.forms.components.*;
import ru.citeck.ecos.files.types.ecos.ui.Form;
import ru.citeck.ecos.index.IndexKey;
import ru.citeck.ecos.index.IndexValue;
import ru.citeck.ecos.index.IndexesService;
import ru.citeck.ecos.index.indexers.EcosDataTypeIndexer;
import ru.citeck.ecos.utils.CommonUtils;
import ru.citeck.ecos.utils.EcosMessages;
import ru.citeck.ecos.utils.EcosPsiUtils;
import ru.citeck.ecos.utils.JsonPrettyPrinter;

import java.util.*;
import java.util.function.Supplier;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class GenerateEcosFormComponent extends SimpleCodeInsightAction {

    public static final String COMPONENTS_PATH = "/components";
    private static final Map<String, Supplier<Component>> COMPONENTS = Map.of(
            SelectJournal.TYPE, SelectJournal::new,
            DateTime.TYPE, DateTime::new,
            CheckBox.TYPE, CheckBox::new,
            TextField.TYPE, TextField::new,
            TextArea.TYPE, TextArea::new,
            Columns.TYPE + " (1 column)", () -> new Columns(1),
            Columns.TYPE + " (2 columns)", () -> new Columns(2),
            Panel.TYPE, Panel::new,
            EcosSelect.TYPE, EcosSelect::new
    );

    private record InsertionPosition(String path, int index) {
    }

    @Override
    public void invoke(@NotNull Project project, @NotNull Editor editor, @NotNull PsiFile psiFile) {

        InsertionPosition insertionPosition = getInsertionPath(editor, psiFile);
        if (insertionPosition == null) {
            return;
        }

        JBPopupFactory
                .getInstance()
                .createPopupChooserBuilder(Arrays.asList(COMPONENTS.keySet().toArray()))
                .setTitle("Select Component Type:")
                .setNamerForFiltering(Object::toString)
                .setRequestFocus(true)
                .setItemChosenCallback(componentType -> onComponentTypeSelected(editor, psiFile, (String) componentType, insertionPosition))
                .createPopup()
                .showInCenterOf(WindowManager.getInstance().getFrame(project).getRootPane());

    }

    private InsertionPosition getInsertionPath(Editor editor, PsiFile psiFile) {

        int selectionStart = editor.getSelectionModel().getSelectionStart();

        PsiElement selectedElement = psiFile.findElementAt(selectionStart);
        if (selectedElement == null) {
            return null;
        }
        PsiElement selectedProperty = PsiTreeUtil.getParentOfType(selectedElement, JsonProperty.class);
        if (selectedProperty == null) {
            return null;
        }

        String qualifiedName = DumbService.getInstance(psiFile.getProject()).computeWithAlternativeResolveEnabled(() ->
                QualifiedNameProviderUtil.getQualifiedName(selectedProperty));

        String path = "/" + qualifiedName
                .replace("]", "/")
                .replace("[", "/")
                .replace(".", "/")
                .replace("//", "/");

        if (!path.contains(COMPONENTS_PATH)) {
            return null;
        }

        path = path.substring(0, path.lastIndexOf(COMPONENTS_PATH) + COMPONENTS_PATH.length());

        int index = 0;

        if (selectedProperty.getChildren().length == 2 && selectedProperty.getChildren()[1] instanceof JsonArray) {
            index = (int) Arrays.stream(selectedProperty.getChildren()[1].getChildren())
                    .map(PsiElement::getTextOffset)
                    .filter(offset -> offset < selectionStart)
                    .count();
        }

        return new InsertionPosition(path, index);

    }

    private void onComponentTypeSelected(Editor editor, @NotNull PsiFile psiFile, String componentType, InsertionPosition insertionPosition) {

        Component component = COMPONENTS.get(componentType).get();

        if (component instanceof InputComponent) {
            JBPopupFactory
                    .getInstance()
                    .createPopupChooserBuilder(getAttributes(psiFile, editor.getProject(), (InputComponent) component))
                    .setTitle("Select Attribute:")
                    .setNamerForFiltering(Object::toString)
                    .setRequestFocus(true)
                    .setItemChosenCallback(attribute -> onAttributeSelected(editor, component, attribute, insertionPosition))
                    .createPopup()
                    .showInCenterOf(WindowManager.getInstance().getFrame(editor.getProject()).getRootPane());
        } else {
            onAttributeSelected(editor, component, "", insertionPosition);
        }
    }

    private void onAttributeSelected(Editor editor, Component component, String attribute, InsertionPosition insertionPosition) {

        Project project = editor.getProject();

        String json;

        try {
            ObjectMapper objectMapper = new ObjectMapper();
            JsonNode jsonNode = objectMapper.readValue(editor.getDocument().getText(), JsonNode.class);

            if (component instanceof InputComponent inputComponent) {
                inputComponent.setKey(attribute.replace(":", "_"));
                inputComponent.getProperties().setAttribute(attribute);
                inputComponent.setLabel(attribute);
            } else {
                Set<String> keys = jsonNode.findValues("key").stream().map(JsonNode::asText).collect(Collectors.toSet());
                String key;
                int i = 0;
                do {
                    key = component.getType() + (++i);
                } while (keys.contains(key));
                component.setKey(key);
                component.setLabel(key);
            }

            ((ArrayNode) jsonNode.at(insertionPosition.path())).insert(
                    insertionPosition.index(),
                    objectMapper.convertValue(component, JsonNode.class)
            );

            json = objectMapper.writer(new JsonPrettyPrinter()).writeValueAsString(jsonNode).replace("\r\n", "\n");

        } catch (Exception ex) {
            EcosMessages.error("Error while insert component", "Incorrect file content", project);
            return;
        }

        String finalJson = json;
        ApplicationManager.getApplication().runWriteAction(
                () -> CommandProcessor.getInstance().runUndoTransparentAction(
                        () -> editor.getDocument().setText(finalJson)
                )
        );


    }

    private List<String> getAttributes(@NotNull PsiFile psiFile, Project project, InputComponent component) {

        IndexesService indexesService = ServiceRegistry.getIndexesService(project);

        List<String> typeRefAttributes = Optional
                .of(psiFile)
                .map(CommonUtils.filterAndCast(JsonFile.class))
                .map(JsonFile::getTopLevelValue)
                .map(CommonUtils.filterAndCast(JsonObject.class))
                .map(jsonObject -> EcosPsiUtils.getProperty(jsonObject, "typeRef"))
                .map(typeRef -> typeRef.replace("emodel/type", "emodel/types-repo"))
                .map(typeRef -> getAttributes(project, typeRef).collect(Collectors.toList()))
                .orElseGet(Collections::emptyList);

        if (!typeRefAttributes.isEmpty()) {
            return typeRefAttributes;
        }

        return component
                .getSupportedArtifactTypes()
                .stream()
                .map(IndexKey::new)
                .flatMap(indexesService::stream)
                .map(IndexValue::getId)
                .collect(Collectors.toCollection(ArrayList::new));

    }

    @NotNull
    private static Stream<String> getAttributes(Project project, String typeRef) {
        IndexesService indexesService = ServiceRegistry.getIndexesService(project);
        Stream<String> attributes = indexesService
                .stream(new IndexKey(typeRef, EcosDataTypeIndexer.ATTRIBUTE))
                .map(IndexValue::getId);

        String parentRef = indexesService
                .stream(new IndexKey(typeRef, "parentRef"))
                .findFirst()
                .map(IndexValue::getId)
                .orElse(null);

        if (parentRef == null) {
            return attributes;
        }

        return Stream.concat(attributes, getAttributes(project, parentRef));
    }

    @Override
    protected void update(@NotNull Presentation presentation, @NotNull Project project, @NotNull Editor editor, @NotNull PsiFile psiFile) {
        boolean isForm = ServiceRegistry.getFileTypeService().isInstance(psiFile, Form.JSON.class);
        presentation.setVisible(isForm && getInsertionPath(editor, psiFile) != null);
    }

}

File: src/main/kotlin/ru/citeck/idea/artifacts/action/codeinsight/GenerateEcosFormComponent.kt
package ru.citeck.idea.artifacts.action.codeinsight

import com.intellij.codeInsight.actions.SimpleCodeInsightAction
import com.intellij.ide.actions.QualifiedNameProviderUtil
import com.intellij.json.psi.JsonArray
import com.intellij.json.psi.JsonProperty
import com.intellij.openapi.actionSystem.Presentation
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.command.CommandProcessor
import com.intellij.openapi.editor.Editor
import com.intellij.openapi.project.DumbService
import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.popup.JBPopupFactory
import com.intellij.openapi.util.ThrowableComputable
import com.intellij.openapi.wm.WindowManager
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiFile
import com.intellij.psi.util.PsiTreeUtil
import ecos.com.fasterxml.jackson210.databind.JsonNode
import ecos.com.fasterxml.jackson210.databind.ObjectMapper
import ecos.com.fasterxml.jackson210.databind.node.ArrayNode
import ru.citeck.ecos.commons.json.Json.mapper
import ru.citeck.idea.artifacts.action.codeinsight.forms.components.*
import ru.citeck.idea.artifacts.ArtifactTypes
import ru.citeck.idea.json.PsiJPath
import ru.citeck.idea.search.index.IndexKey
import ru.citeck.idea.search.index.IndexValue
import ru.citeck.idea.search.index.IndexesService
import ru.citeck.idea.utils.CiteckMessages
import java.util.*
import java.util.function.Supplier
import java.util.stream.Collectors

class GenerateEcosFormComponent : SimpleCodeInsightAction() {

    companion object {
        const val COMPONENTS_PATH: String = "/components"

        private val COMPONENTS: Map<String, Supplier<Component>> = mapOf(
            SelectJournal.TYPE to Supplier { SelectJournal() },
            DateTime.TYPE to Supplier { DateTime() },
            CheckBox.TYPE to Supplier { CheckBox() },
            TextField.TYPE to Supplier { TextField() },
            TextArea.TYPE to Supplier { TextArea() },
            Columns.TYPE + " (1 column)" to Supplier {
                Columns(
                    1
                )
            },
            Columns.TYPE + " (2 columns)" to Supplier {
                Columns(
                    2
                )
            },
            Panel.TYPE to Supplier { Panel() },
            EcosSelect.TYPE to Supplier { EcosSelect() }
        )
    }

    override fun invoke(project: Project, editor: Editor, psiFile: PsiFile) {
        val insertionPosition = getInsertionPath(editor, psiFile) ?: return

        JBPopupFactory
            .getInstance()
            .createPopupChooserBuilder(listOf<Any>(*COMPONENTS.keys.toTypedArray()))
            .setTitle("Select Component Type:")
            .setNamerForFiltering { obj: Any -> obj.toString() }
            .setRequestFocus(true)
            .setItemChosenCallback { componentType: Any? ->
                onComponentTypeSelected(
                    editor,
                    psiFile,
                    componentType as String?,
                    insertionPosition
                )
            }
            .createPopup()
            .showInCenterOf(WindowManager.getInstance().getFrame(project)!!.rootPane)
    }

    private fun getInsertionPath(editor: Editor, psiFile: PsiFile): InsertionPosition? {
        val selectionStart = editor.selectionModel.selectionStart

        val selectedElement = psiFile.findElementAt(selectionStart) ?: return null
        val selectedProperty: PsiElement? = PsiTreeUtil.getParentOfType(
            selectedElement,
            JsonProperty::class.java
        )
        if (selectedProperty == null) {
            return null
        }

        val qualifiedName: String = DumbService.getInstance(psiFile.project)
            .computeWithAlternativeResolveEnabled(
                ThrowableComputable<String, RuntimeException> {
                    QualifiedNameProviderUtil.getQualifiedName(selectedProperty)
                }
            )

        var path = "/" + qualifiedName
            .replace("]", "/")
            .replace("[", "/")
            .replace(".", "/")
            .replace("//", "/")

        if (!path.contains(COMPONENTS_PATH)) {
            return null
        }

        path = path.substring(0, path.lastIndexOf(COMPONENTS_PATH) + COMPONENTS_PATH.length)

        var index = 0

        if (selectedProperty.children.size == 2 && selectedProperty.children[1] is JsonArray) {
            index = Arrays.stream(selectedProperty.children[1].children)
                .map { obj: PsiElement -> obj.textOffset }
                .filter { offset: Int -> offset < selectionStart }
                .count().toInt()
        }

        return InsertionPosition(path, index)
    }

    private fun onComponentTypeSelected(
        editor: Editor,
        psiFile: PsiFile,
        componentType: String?,
        insertionPosition: InsertionPosition
    ) {
        val component = COMPONENTS[componentType]!!
            .get()

        if (component is InputComponent) {
            JBPopupFactory
                .getInstance()
                .createPopupChooserBuilder(getAttributes(psiFile, editor.project!!, component))
                .setTitle("Select Attribute:")
                .setNamerForFiltering { obj: String -> obj }
                .setRequestFocus(true)
                .setItemChosenCallback { attribute: String ->
                    onAttributeSelected(
                        editor,
                        component,
                        attribute,
                        insertionPosition
                    )
                }
                .createPopup()
                .showInCenterOf(WindowManager.getInstance().getFrame(editor.project)!!.rootPane)
        } else {
            onAttributeSelected(editor, component, "", insertionPosition)
        }
    }

    private fun onAttributeSelected(
        editor: Editor,
        component: Component,
        attribute: String,
        insertionPosition: InsertionPosition
    ) {
        val project = editor.project

        val json: String

        try {
            val objectMapper = ObjectMapper()
            val jsonNode = objectMapper.readValue(
                editor.document.text,
                JsonNode::class.java
            )

            if (component is InputComponent) {
                component.setKey(attribute.replace(":", "_"))
                component.properties.attribute = attribute
                component.setLabel(attribute)
            } else {
                val keys = jsonNode.findValues("key")
                    .stream()
                    .map { obj: JsonNode -> obj.asText() }
                    .collect(Collectors.toSet())
                var key: String
                var i = 0
                do {
                    key = component.getType() + (++i)
                } while (keys.contains(key))
                component.setKey(key)
                component.setLabel(key)
            }

            (jsonNode.at(insertionPosition.path) as ArrayNode).insert(
                insertionPosition.index,
                objectMapper.convertValue(component, JsonNode::class.java)
            )

            json = mapper.toPrettyStringNotNull(jsonNode)
        } catch (ex: Exception) {
            CiteckMessages.error("Error while insert component", "Incorrect file content", project)
            return
        }

        ApplicationManager.getApplication().runWriteAction {
            CommandProcessor.getInstance().runUndoTransparentAction {
                editor.document.setText(json)
            }
        }
    }

    private fun getAttributes(psiFile: PsiFile, project: Project, component: InputComponent): List<String> {

        val indexesService = IndexesService.getInstance(project)

        val typeRef = PsiJPath.parse("$.typeRef").getStr(psiFile) ?: return emptyList()
        val typeElement = indexesService.stream(IndexKey(typeRef)).findFirst().orElse(null)

        val typeRefFile = typeElement.getPsiElement(project) ?: return emptyList()
        val typeData = mapper.readData(typeRefFile.text)

        val typeRefAtts = typeData["$.model.attributes[:].id"].asStrList()

        if (typeRefAtts.isNotEmpty()) {
            return typeRefAtts
        }

        return component
            .getSupportedArtifactTypes()
            .stream()
            .map { type: String? -> IndexKey(type!!) }
            .flatMap { key: IndexKey? -> indexesService.stream(key!!) }
            .map(IndexValue::id)
            .collect(Collectors.toCollection { ArrayList() })
    }

    override fun update(presentation: Presentation, project: Project, editor: Editor, psiFile: PsiFile) {
        presentation.isVisible = ArtifactTypes.isForm(psiFile) && getInsertionPath(editor, psiFile) != null
    }

    private data class InsertionPosition(val path: String, val index: Int)
}
